tsne/bh_sne_src/tsne.cpp

```diff
@@ -83,427 +124,427 @@ void TSNE::run(double* X, int N, int D, double* Y, int no_dims, double perplexit
-    unsigned int* row_P = *_row_P;
-    unsigned int* col_P = *_col_P;
-    double* val_P = *_val_P;
-    double* cur_P = (double*) malloc((N - 1) * sizeof(double));
-    if(cur_P == NULL) { fprintf(stderr,"Memory allocation failed!\n"); exit(1); }
+    _row_P->resize(N + 1);
+    _col_P->resize(N * K, 0);
+    _val_P->resize(N * K, 0);
+    vector<unsigned int>& row_P = *_row_P;
+    vector<unsigned int>& col_P = *_col_P;
+    vector<double>& val_P = *_val_P;
+    vector<double> cur_P(N - 1);
     row_P[0] = 0;
-    for(int n = 0; n < N; n++) row_P[n + 1] = row_P[n] + (unsigned int) K;
+    for (int n = 0; n < N; n++) row_P[n + 1] = row_P[n] + (unsigned int)K;
 
     // Build ball tree on data set
-    VpTree<DataPoint, euclidean_distance>* tree = new VpTree<DataPoint, euclidean_distance>();
-    vector<DataPoint> obj_X(N, DataPoint(D, -1, X));
-    for(int n = 0; n < N; n++) obj_X[n] = DataPoint(D, n, X + n * D);
-    tree->create(obj_X);
+    VpTree<DataPoint, euclidean_distance> tree((euclidean_distance(D)));
+    vector<DataPoint> obj_X(N);
+    for (int n = 0; n < N; n++) obj_X[n] = DataPoint(X + n * D);
+    tree.create(obj_X);
 
     // Loop over all points to find nearest neighbors
-    fprintf(stderr,"Building tree...\n");
+    fprintf(stderr, "Building tree...\n");
     vector<DataPoint> indices;
     vector<double> distances;
-    for(int n = 0; n < N; n++) {
-
-        if(n % 10000 == 0) fprintf(stderr," - point %d of %d\n", n, N);
+    for (int n = 0; n < N; n++) {
+        if (n % 10000 == 0)
+            fprintf(stderr, " - point %d of %d\n", n, N);
 
         // Find nearest neighbors
         indices.clear();
         distances.clear();
-        tree->search(obj_X[n], K + 1, &indices, &distances);
+        tree.search(obj_X[n], K + 1, &indices, &distances);
 
         // Initialize some variables for binary search
         bool found = false;
         double beta = 1.0;
         double min_beta = -DBL_MAX;
-        double max_beta =  DBL_MAX;
+        double max_beta = DBL_MAX;
         double tol = 1e-5;
 
         // Iterate until we found a good perplexity
-        int iter = 0; double sum_P;
-        while(!found && iter < 200) {
-
+        int iter = 0;
+        double sum_P;
+        while (!found && iter < 200) {
             // Compute Gaussian kernel row
-            for(int m = 0; m < K; m++) cur_P[m] = exp(-beta * distances[m + 1] * distances[m + 1]);
+            for (int m = 0; m < K; m++)
+                cur_P[m] = exp(-beta * distances[m + 1] * distances[m + 1]);
 
             // Compute entropy of current row
             sum_P = DBL_MIN;
-            for(int m = 0; m < K; m++) sum_P += cur_P[m];
+            for (int m = 0; m < K; m++) sum_P += cur_P[m];
             double H = .0;
-            for(int m = 0; m < K; m++) H += beta * (distances[m + 1] * distances[m + 1] * cur_P[m]);
+            for (int m = 0; m < K; m++)
+                H += beta * (distances[m + 1] * distances[m + 1] * cur_P[m]);
```

Comment:
You could do the same factorization of beta here.

Reply:
Yeah, there's more that can be done, but 57% of profile samples are on line 264 of sptree.cpp so unless something was offensive to my sense of aesthetics I didn't look too closely.  Will update this, however.